new-build command for creating BuildConfigs

soltysh opened the pull request:
This should fix #2695.
@smarterclayton @csrwng ptal

This PR also includes a fix (2nd commit) for the case when generating new-app/new-build from repo containing Dockerfile (eg. https://github.com/openshift/ruby-hello-world) with forced build type (`--build=source`), previously that always created dockerStrategy.

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -160,10 +203,10 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
+	return nil
+}
 
+func produceOutput(f *clientcmd.Factory, out io.Writer, result *newcmd.AppResult) error {
```
I think something like 'createObjects' would be a more appropriate name. produceOutput sounds like you're printing something.

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -132,20 +185,10 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
+	return nil
+}
 
+func parseLabels(c *cobra.Command, result *newcmd.AppResult) error {
```
applyLabels or setLabels

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
+	return nil
+}
+
+func parseParameters(f *clientcmd.Factory, c *cobra.Command, args []string, config *newcmd.AppConfig) error {
```
setupAppConfig

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
+				err = errs.Errors()[0]
+			}
+		}
+		if err == newcmd.ErrNoInputs {
```
this should be after the enclosing if

soltysh replied:
What's wrong with this being here, at this point you are certain that the `err != nil` and you're just checking its type?

csrwng replied:
sorry, yup makes sense.... for some reason I thought it was inside the previous if

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+)
+
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code, and/or images.
```
...source code, and/or images. -> source code.

(it's either source code or source code + image)

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code, and/or images.
+
+This command will try to create a BuildConfig for your application using images, or code that has
```
...using images, or code -> ...using images and code

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+a public repository located on your system. It will lookup the images on the
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
```
s/has/has a/

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	newBuildLong = `Create a new build in OpenShift by specifying source code, and/or images.
+
+This command will try to create a BuildConfig for your application using images, or code that has
+a public repository located on your system. It will lookup the images on the
```
not necessarily located on your system... you can give it a URL

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+it into an image that can run inside of a pod. Local source must be in a git repository that has
+remote repository that the OpenShift instance can see.
+
+Once the BuildConfig will be created you may need to run a build with 'start-build'.`
```
BuildConfig will be -> BuildConfig is

csrwng replied:
s/BuildConfig/build configuration/

soltysh replied:
Throughout all the logs/msgs we keep original object name, thus here I'll leave BuildConfig.

soltysh replied:
It's easier to find in docs BuildConfig than build configuration.

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	newBuildExample = `  // Create a BuildConfig based on the source code in the current git repository (with a public remote) and a Docker image
+  $ %[1]s new-build . --docker-image=repo/langimage
+
+  // Create a NodeJS application based on the provided [image]~[source code] combination
```
s/application/build configuration/

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+
+Once the BuildConfig will be created you may need to run a build with 'start-build'.`
+
+	newBuildExample = `  // Create a BuildConfig based on the source code in the current git repository (with a public remote) and a Docker image
```
s/BuildConfig/build configuration/

soltysh replied:
Same as above, leaving BuildConfig.

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+  // Create a NodeJS application based on the provided [image]~[source code] combination
+  $ %[1]s new-build openshift/nodejs-010-centos7~https://bitbucket.com/user/nodejs-app
+
+  // Create an application from a remote repository using its beta2 branch
```
s/application/build configuration/

soltysh replied:
Will be BuildConfig.

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -32,12 +32,12 @@ const (
-it into an image that can run inside of a pod. Local source must be in a git repository that has 
-remote repository that the OpenShift instance can see. The images will be deployed via a deployment 
-configuration, and a service will be hooked up to the first public port of the app. You may either specify 
+or code that has a public repository located on your system. It will lookup the images on the
```
I would remove 'located on your system.'

soltysh replied:
Done in both newapp.go and newbuild.go

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -32,12 +32,12 @@ const (
+or code that has a public repository located on your system. It will lookup the images on the
+local Docker installation (if available), a Docker registry, or an OpenShift image stream.
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
```
s/has/has a/

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -32,12 +32,12 @@ const (
+If you specify a source code URL, it will set up a build that takes your source code and converts
+it into an image that can run inside of a pod. Local source must be in a git repository that has
+remote repository that the OpenShift instance can see. The images will be deployed via a deployment
+configuration, and a service will be hooked up to the first public port of the app. You may either specify
```
s/hooked up/connected/

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	config := newcmd.NewAppConfig(typer)
+
+	cmd := &cobra.Command{
+		Use:     "new-build (IMAGE | IMAGESTREAM | TEMPLATE | PATH | URL ...)",
```
TEMPLATE should not be in the list

csrwng replied:
IMAGE and IMAGESTREAM should be there only if associated with SOURCE

smarterclayton replied:
It's ok to have "IMAGE SOURCE" - the normal logic should associate them.

----- Original Message -----
> > +  $ %[1]s new-build . --docker-image=repo/langimage
> > +
> > +  // Create a NodeJS application based on the provided [image]~[source
> > code] combination
> > +  $ %[1]s new-build
> > openshift/nodejs-010-centos7~https://bitbucket.com/user/nodejs-app
> > +
> > +  // Create an application from a remote repository using its beta2 branch
> > +  $ %[1]s new-build https://github.com/openshift/ruby-hello-world#beta2`
> > +)
> > +
> > +// NewCmdNewBuild implements the OpenShift cli new-build command
> > +func NewCmdNewBuild(fullName string, f *clientcmd.Factory, out io.Writer)
> > *cobra.Command {
> > +	_, typer := f.Object()
> > +	config := newcmd.NewAppConfig(typer)
> > +
> > +	cmd := &cobra.Command{
> > +		Use:     "new-build (IMAGE | IMAGESTREAM | TEMPLATE | PATH | URL ...)",
> 
> IMAGE and IMAGESTREAM should be there only if associated with SOURCE
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2747/files#r31661607

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+
+	cmd := &cobra.Command{
+		Use:     "new-build (IMAGE | IMAGESTREAM | TEMPLATE | PATH | URL ...)",
+		Short:   "Create a new BuildConfig",
```
s/BuildConfig/build configuration/

soltysh replied:
Same as above, leaving BuildConfig for clarity.

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+		},
+	}
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
```
"to build this application" -> "in the build configuration"

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
```
"to use in the app" -> "to use as a builder"

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
```
"to include in the app" -> "to use as a builder"

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
+	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated application artifacts")
```
s/application artifacts/build artifacts"

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
+	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated application artifacts")
+	cmd.Flags().StringVar(&config.TypeOfBuild, "build", "", "Specify the type of build to use if you don't want to detect (docker|source).")
+	cmd.Flags().StringP("labels", "l", "", "Label to set in all resources for this application.")
```
... to set in all generated resources.

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml.") //|template|templatefile.")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
```
Remove

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+	cmd.Flags().StringP("output", "o", "", "Output format. One of: json|yaml.") //|template|templatefile.")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().Bool("no-headers", false, "When using the default output, don't print headers.")
+	cmd.Flags().String("template", "", "Template string or path to template file to use when -o=template or -o=templatefile.  The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview]")
```
Remove - Unless we want to support templates in new-build

soltysh replied:
I'd leave it, without it you cannot use -o json|yaml which is useful if you just want to see the objects created. The printer internally reaches for template, without it you'll get an error that you're trying to access flag that wasn't defined, been there and tested that.

smarterclayton replied:
We don't want to support real templates - you can go back to using the AddPrinterFlags function

csrwng commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,111 @@
+		}
+		if err == newcmd.ErrNoInputs {
+			// TODO: suggest things to the user
+			return cmdutil.UsageError(c, "You must specify one or more images, image streams or source code locations to create a BuildConfig.")
```
or source code locations -> and source code locations

csrwng commented on pkg/generate/app/cmd/newapp.go:
```diff
@@ -418,10 +418,21 @@ type AppResult struct {
+}
+
+// run executes the provided config applying provided acceptors.
+func (c *AppConfig) run(out io.Writer, acceptors app.Acceptors) (*AppResult, error) {
```
Line 444 - if you asked for NewBuild and you don't have source you should return an error.

csrwng commented on pkg/generate/app/pipeline.go:
```diff
@@ -311,6 +311,30 @@ func objectMetaData(raw interface{}) (runtime.Object, *kapi.ObjectMeta, error) {
+	if err != nil {
+		return false
+	}
+	return kind == "BuildConfig"
```
What about the output image stream? or the input image stream if creating from a docker repo?

smarterclayton replied:
Always end an output image stream from new-build, unless we support an argument that lets you push to an external Docker repo (nice to have, can be post 3.0)

----- Original Message -----
> > @@ -311,6 +311,30 @@ func objectMetaData(raw interface{}) (runtime.Object,
> > *kapi.ObjectMeta, error) {
> >  	return obj, meta, nil
> >  }
> >  
> > +type acceptBuildConfigs struct {
> > +	typer runtime.ObjectTyper
> > +}
> > +
> > +// Accept accepts just BuildConfigs
> > +func (a *acceptBuildConfigs) Accept(from interface{}) bool {
> > +	obj, _, err := objectMetaData(from)
> > +	if err != nil {
> > +		return false
> > +	}
> > +	_, kind, err := a.typer.ObjectVersionAndKind(obj)
> > +	if err != nil {
> > +		return false
> > +	}
> > +	return kind == "BuildConfig"
> 
> What about the output image stream? or the input image stream if creating
> from a docker repo?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2747/files#r31661250

soltysh replied:
We do support pushes to external repo -> DockerImage.

smarterclayton replied:
It's probably not going to be the default - a user should have to give us a --to-docker flag (or similar).  Ideally I'd like to reserve "--to" to be an Openshift image tag (so you can easily spin up a new build pointing to that tag, vs using a default name).

> On Jun 3, 2015, at 6:04 PM, Maciej Szulik <notifications@github.com> wrote:
> 
> In pkg/generate/app/pipeline.go:
> 
> > @@ -311,6 +311,30 @@ func objectMetaData(raw interface{}) (runtime.Object, *kapi.ObjectMeta, error) {
> >  	return obj, meta, nil
> >  }
> >  
> > +type acceptBuildConfigs struct {
> > +	typer runtime.ObjectTyper
> > +}
> > +
> > +// Accept accepts just BuildConfigs
> > +func (a *acceptBuildConfigs) Accept(from interface{}) bool {
> > +	obj, _, err := objectMetaData(from)
> > +	if err != nil {
> > +		return false
> > +	}
> > +	_, kind, err := a.typer.ObjectVersionAndKind(obj)
> > +	if err != nil {
> > +		return false
> > +	}
> > +	return kind == "BuildConfig"
> We do support pushes to external repo -> DockerImage.
> 
> —
> Reply to this email directly or view it on GitHub.
>

soltysh replied:
I'll catch you tomorrow on IRC to talk this through.

csrwng commented:
I don't see a need for new-build to handle templates, since you can already do that with new-app, plus only creating buildConfigs from a template might not make sense. If that's the case, you need to add code in new-app.go to not allow templates when running in new-build mode.

smarterclayton commented:
Agreed

----- Original Message -----
> I don't see a need for new-build to handle templates, since you can already
> do that with new-app, plus only creating buildConfigs from a template might
> not make sense. If that's the case, you need to add code in new-app.go to
> not allow templates when running in new-build mode.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2747#issuecomment-108591283

soltysh commented:
If we remove templates, I need to remove --output as well, see [this comment](https://github.com/openshift/origin/pull/2747#discussion_r31675254)

soltysh commented:
@csrwng addressed most of your comments for others I wrote my reasoning. I've also fixed the tests that was failing. The only left question is whether we want `--output` and `--templates` flags.

smarterclayton commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
+		}
+		if err == newcmd.ErrNoInputs {
+			// TODO: suggest things to the user
+			return cmdutil.UsageError(c, "You must specify one or more images, image streams, templates or source code locations to create an application.")
```
Remove templates

smarterclayton replied:
Oh, this is the new code

soltysh replied:
Yes, this is part of newapp.go, so this should be as is.

soltysh commented:
@smarterclayton I've changed newbuild.go to use `AddPrinterFlags`. The only thing left will be then `--to-docker` flag I'm hoping to discuss with you tomorrow.

mfojtik commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -106,6 +106,59 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
+			}
+			fmt.Fprintf(c.Out(), "Service %q created at %s%s\n", t.Name, t.Spec.PortalIP, portMappings)
+		case *buildapi.BuildConfig:
+			fmt.Fprintf(c.Out(), "A build was created - you can run `osc start-build %s` to start it.\n", t.Name)
```
don't we have constant for `osc` string?

soltysh replied:
Fixed.

mfojtik commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,107 @@
+
+	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code in the build configuration.")
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to to use as a builder.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to use as a builder.")
```
DockerImages vs. 'docker-image'

soltysh replied:
Internally DockerImages allows having more than one.

mfojtik commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,107 @@
+	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to to use as a builder.")
+	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to use as a builder.")
+	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated build artifacts")
+	cmd.Flags().StringVar(&config.TypeOfBuild, "build", "", "Specify the type of build to use if you don't want to detect (docker|source).")
```
build-type ?

soltysh replied:
Agreed on IRC to rename that to strategy, to match the produced output.

smarterclayton commented:
We want --output and --templates like osc new-app yes - all our "generate" commands should be able to spit out yaml so that someone can play around and try stuff.

----- Original Message -----
> @csrwng addressed most of your comments for others I wrote my reasoning. I've
> also fixed the tests that was failing. The only left question is whether we
> want `--output` and `--templates` flags.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/2747#issuecomment-108634693

soltysh commented:
@smarterclayton added `--to-docker` flag.
@csrwng I've renamed `--build` to `--strategy` as discussed on IRC.
ptal
[test] in the meantime

smarterclayton commented on pkg/cmd/cli/cmd/newbuild.go:
```diff
@@ -0,0 +1,109 @@
+const (
+	newBuildLong = `Create a new build in OpenShift by specifying source code.
+
+This command will try to create a BuildConfig for your application using images and
```
`build configuration (BuildConfig)`

soltysh replied:
Fixed.

soltysh commented:
Updated and fixed test error.

soltysh commented:
I don't get the error, it's not related to this PR.
re-[test]

soltysh commented:
Again, rebased and waiting for tests result.

smarterclayton commented:
LGTM [merge]
